paasta_tools/setup_smartstack_namespace_k8s_svcs.py

```diff
@@ -0,0 +1,579 @@
+    """Build a Kubernetes Service object for a smartstack namespace.
+
+    Args:
+        kube_namespace: Kubernetes namespace to create service in
+        smartstack_namespace: Smartstack namespace (e.g. "service.main")
+
+    Returns:
+        V1Service object ready to be applied
+    """
+    service_name = sanitise_kubernetes_service_name(smartstack_namespace)
+    metadata = k8s.V1ObjectMeta(
+        name=service_name,
+        namespace=kube_namespace,
+        labels=SERVICE_LABELS.copy(),
+        annotations=SERVICE_ANNOTATIONS.copy(),
+    )
+    port_spec = k8s.V1ServicePort(
+        name="application",
+        port=80,  # Required by k8s API but not used for headless services
+        target_port=CONTAINER_PORT_NAME,
+        protocol="TCP",
+        app_protocol="http",
+    )
+    spec = k8s.V1ServiceSpec(
+        cluster_ip="None",  # Headless service - no virtual IP, just DNS
+        selector={registration_label(smartstack_namespace): "true"},
+        ports=[port_spec],
+    )
+    return k8s.V1Service(
+        api_version="v1",
+        kind="Service",
+        metadata=metadata,
+        spec=spec,
+    )
+
+
+def server_side_apply_service(
+    kube_client: KubeClient,
+    kube_namespace: str,
+    service: k8s.V1Service,
+    dry_run: bool = False,
+) -> None:
+    log.debug(f"Reconciling k8s Service {kube_namespace}/{service.metadata.name}")
+    serialized = kube_client.core.api_client.sanitize_for_serialization(service)
+
+    patch_kwargs = dict(
+        name=service.metadata.name,
+        namespace=kube_namespace,
+        body=serialized,
+        field_manager=FIELD_MANAGER,
+        force=True,
+        content_type="application/apply-patch+yaml",
+    )
+    if dry_run:
+        patch_kwargs["dry_run"] = "All"
+
+    try:
+        kube_client.core.patch_namespaced_service(**patch_kwargs)
+    except ApiException as e:
+        if e.status != 404:
+            raise
+
+        create_kwargs = dict(namespace=kube_namespace, body=service)
+        if dry_run:
+            create_kwargs["dry_run"] = "All"
+
+        created = False
+        try:
+            kube_client.core.create_namespaced_service(**create_kwargs)
+            created = True
+            if not dry_run:
+                kube_client.core.patch_namespaced_service(**patch_kwargs)
+        except Exception:
+            if created and not dry_run:
+                log.warning(
+                    f"Rolling back Service {kube_namespace}/{service.metadata.name} after apply error"
+                )
+                try:
+                    kube_client.core.delete_namespaced_service(
+                        name=service.metadata.name,
+                        namespace=kube_namespace,
+                    )
+                except Exception:
+                    log.exception(
+                        f"Failed to rollback Service {kube_namespace}/{service.metadata.name} after apply error"
+                    )
+            raise
```

Comment:
hmm, I wonder if we should just create the Service as we'd want it to look over Create+Patch?

(i think that's the pattern we take in skj and whatnot - 3 different paths for create vs modify vs delete)